src/storage-client/src/source/persist_source.rs

```diff
@@ -528,18 +588,270 @@ impl PersistSourceDataStats<'_> {
+                    (time, part)
+                } else {
+                    match data_input.next().await {
+                        Some(Either::Right((time, part))) => {
+                            // Downgrade the output frontier to this part's time. This is useful
+                            // "close" timestamp's from previous parts, even if we don't yet
+                            // emit this part. Note that this is safe because `data_input` ensures
+                            // time-ordering.
+                            output_frontier = Antichain::from_elem(time.clone());
+                            cap_set.downgrade(output_frontier.iter());
+
+                            // If the most recent value's time is _beyond_ the
+                            // `flow_control` frontier (which takes into account the `summary`), we
+                            // have emitted an entire `summary` worth of data, and can store this
+                            // value for later.
+                            if inflight_bytes >= flow_control_max_bytes
+                                && !PartialOrder::less_than(
+                                    &output_frontier,
+                                    &flow_control_frontier,
+                                )
+                            {
+                                pending_parts.push_back((time, part));
+                                continue 'emitting_parts;
+                            }
+                            (time, part)
+                        }
+                        Some(Either::Left(prog)) => {
+                            output_frontier = prog;
+                            cap_set.downgrade(output_frontier.iter());
+                            continue 'emitting_parts;
+                        }
+                        None => {
+                            if pending_parts.is_empty() {
+                                break 'emitting_parts;
+                            } else {
+                                continue 'emitting_parts;
+                            }
+                        }
+                    }
+                };
+
+                // Store the value with the _frontier_ the `flow_control_input` must reach
+                // to retire it. Note that if this `results_in` is `None`, then we
+                // are at `T::MAX`, and give up on flow_control entirely.
+                //
+                // SUBTLE: If we stop storing these parts, we will likely never check the
+                // `flow_control_input` ever again. This won't pile up data as that input
+                // only has frontier updates. There may be spurious activations from it though.
+                //
+                // Also note that we don't attempt to handle overflowing the `u64` part counter.
+                if let Some(emission_ts) = flow_control.summary.results_in(&time) {
```

Comment:
It sounds simpler to simply unwrap the `results_in` call. I don't think having to reason about subtle behavior at `u64::MAX` is worth it since we certainly don't expect to ever reach this number of parts

Reply:
This is actually important if we are at the empty frontier, I believe!

Reply:
You are processing a specific time here so it can't be the empty frontier